### PR TITLE
[Merged by Bors] - feat: port Data.Num.Lemmas

### PR DESCRIPTION
**Pull request by Komyyy**
---

* [`data.num.lemmas`@`7ac96dea0572f31067e9e35bebbecef7fa3fa91a`..`2196ab363eb097c008d4497125e0dde23fb36db2`](https://leanprover-community.github.io/mathlib-port-status/file/data/num/lemmas?range=7ac96dea0572f31067e9e35bebbecef7fa3fa91a..2196ab363eb097c008d4497125e0dde23fb36db2)
- [x] depends on: #2780
- [x] depends on: #2781
- [x] depends on: #2782
- [x] depends on: #2783

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2780
* leanprover-community/mathlib4#2781
* leanprover-community/mathlib4#2782
* leanprover-community/mathlib4#2783
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Data/Num/Lemmas.lean [hunk 1]
```diff
@@ -0,0 +1,1819 @@
...
+theorem zneg_toZNumNeg (n : Num) : -n.toZNumNeg = n.toZNum := by cases n <;> rfl
+#align num.zneg_to_znum_neg Num.zneg_toZNumNeg
+
+theorem toZNum_inj {m n : Num} : m.toZNum = n.toZNum ↔ m = n :=
+  ⟨fun h => by cases m <;> cases n <;> cases h <;> rfl, congr_arg _⟩
+#align num.to_znum_inj Num.toZNum_inj
+
+
+@[simp] -- @[norm_cast squash]
+        -- Porting note: `norm_cast squash` requires no `↑` on rhs in Nathlib4.
```
Comment (semorrison) [inline Mathlib/Data/Num/Lemmas.lean:785]:
```suggestion
@[simp] -- @[norm_cast squash]
        -- Porting note: `norm_cast squash` requires no `↑` on rhs in mathlib4.
```
So do we need to do something besides just remove the attribute?

File: Mathlib/Data/Num/Lemmas.lean [hunk 2]
```diff
@@ -0,0 +1,1819 @@
...
+variable {α : Type _}
+
+theorem cast_to_znum : ∀ n : PosNum, (n : ZNum) = ZNum.pos n
+  | 1 => rfl
+  | bit0 p => (ZNum.bit0_of_bit0 p).trans <| congr_arg _ (cast_to_znum p)
+  | bit1 p => (ZNum.bit1_of_bit1 p).trans <| congr_arg _ (cast_to_znum p)
+#align pos_num.cast_to_znum PosNum.cast_to_znum
+
+-- attribute [-simp] Int.add_neg_one -- Porting note: `Int.add_neg_one` doesn't have this.
+
```
Comment (semorrison) [inline Mathlib/Data/Num/Lemmas.lean:1185]:
```suggestion
```
Just remove.

File: Mathlib/Data/Num/Lemmas.lean [hunk 3]
```diff
@@ -0,0 +1,1819 @@
...
+        apply mul_pos
+    zero_le_one := by decide }
+#align znum.linear_ordered_comm_ring ZNum.linearOrderedCommRing
+
+@[simp, norm_cast]
+theorem cast_sub [Ring α] (m n) : ((m - n : ZNum) : α) = m - n := by simp [sub_eq_neg_add]
+#align znum.cast_sub ZNum.cast_sub
+
+-- Porting note: Required for the notation `-[n+1]`.
+open Int in
```
Comment (semorrison) [inline Mathlib/Data/Num/Lemmas.lean:1533]: Can you just open `Int` for the whole file, or section, or something? I don't think this needs repeating so many times.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**